### PR TITLE
Add logging

### PR DESCRIPTION
**Pull request by js6pak**
- Logging to folder different for each port
- Compressing old log files
- Cleanup console output (print only hour, log full date)
- Log SCP:SL process output

*This PR is much more like proposition/draft than actual PR, but it's working fine.*

### CODE REVIEW

**Comment by MichalPetryka** [comment]
System.IO.Compression doesn't work in CoreRT

**Comment by js6pak** [comment]
Why not? It is nuget package now in .net core, I tested it and it works both on linux and windows.

**Comment by iRebbok** [comment]
Btw since WSL 2 uses linux kernel, you could test if it works. This will help us make sure it works well.

File: LocalAdmin V2.csproj [hunk 1]
```diff
@@ -47,5 +47,6 @@
     </ItemGroup>
     <ItemGroup>
         <PackageReference Include="Microsoft.DotNet.ILCompiler" Version="1.0.0-alpha-*" />
+        <PackageReference Include="System.IO.Compression" Version="4.3.0" />
```
Comment (MichalPetryka) [inline LocalAdmin V2.csproj:50]: Redundant

File: IO/Logging/Logger.cs [hunk 2]
```diff
@@ -1,29 +1,53 @@
...
-            using (var streamWriter = new StreamWriter("localadmin.log"))
-            {
-                streamWriter.WriteLine(File.ReadAllText("localadmin.log"));
-                streamWriter.WriteLine(obj.ToString());
-            }
+            var content = obj.ToString();
+            content = string.IsNullOrEmpty(content) ? string.Empty : $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss.fff zzz}] {content}";
+
+            Writer.Write(content);
+            Writer.Flush();
```
Comment (MichalPetryka) [inline IO/Logging/Logger.cs:45]: No reason to flush here

File: IO/Logging/Logger.cs [hunk 3]
```diff
@@ -1,29 +1,53 @@
...
+                    originalFileStream.CopyTo(compressionStream);
+                }
+
+                File.Delete(filePath);
             }
+
+            Writer = File.AppendText(filePath);
         }
 
         public static void Log(object obj)
```
Comment (MichalPetryka) [inline IO/Logging/Logger.cs:36]: I'd make this take a string instead

File: IO/Logging/Logger.cs [hunk 4]
```diff
@@ -1,29 +1,53 @@
...
+
+        public static void Setup(ushort port)
         {
-            if (!File.Exists("localadmin.log")) File.Create("localadmin.log");
+            var directoryPath = Path.Combine("logs", port.ToString());
+            var filePath = Path.Combine(directoryPath, "latest.log");
 
-            using (var streamWriter = new StreamWriter("localadmin.log"))
+            Directory.CreateDirectory(directoryPath);
+            if (File.Exists(filePath))
```
Comment (MichalPetryka) [inline IO/Logging/Logger.cs:17]: Just use File.Move instead, compression shouldn't be done here

File: IO/Logging/Logger.cs [hunk 5]
```diff
@@ -1,29 +1,53 @@
...
+                    using var compressedFileStream = File.Create(Path.Combine(directoryPath, $"{creationTime}-{count}.log.gz"));
+                    using var compressionStream = new GZipStream(compressedFileStream, CompressionMode.Compress);
+
+                    originalFileStream.CopyTo(compressionStream);
+                }
+
+                File.Delete(filePath);
             }
+
+            Writer = File.AppendText(filePath);
```
Comment (MichalPetryka) [inline IO/Logging/Logger.cs:33]: Consider compressing here instead

File: IO/Logging/Logger.cs [hunk 6]
```diff
@@ -1,29 +1,53 @@
...
+using System.IO;
+using System.IO.Compression;
+
+namespace LocalAdmin.V2.IO.Logging
 {
-    /*
     public static class Logger
     {
-        public static void Log(string text)
+        public static TextWriter? Writer { get; private set; }
```
Comment (MichalPetryka) [inline IO/Logging/Logger.cs:9]: Should be disposed on process exit, could be changed to StreamWriter instead

File: IO/Logging/Logger.cs [hunk 7]
```diff
@@ -1,29 +1,53 @@
...
+            var content = obj.ToString();
+            content = string.IsNullOrEmpty(content) ? string.Empty : $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss.fff zzz}] {content}";
+
+            Writer.Write(content);
+            Writer.Flush();
+        }
+
+        public static void LogLine(object obj)
+        {
+            Log(obj + Environment.NewLine);
```
Comment (MichalPetryka) [inline IO/Logging/Logger.cs:50]: I'd consider using WriteLine instead of Write text + newline in case the implementation differs on some platforms

File: IO/Logging/Logger.cs [hunk 8]
```diff
@@ -1,29 +1,53 @@
-﻿namespace LocalAdmin.V2.IO.Logging
+﻿using System;
+using System.IO;
+using System.IO.Compression;
+
+namespace LocalAdmin.V2.IO.Logging
 {
-    /*
     public static class Logger
```
Comment (MichalPetryka) [inline IO/Logging/Logger.cs:7]: This class should use locks in order to be threadsafe

File: IO/ConsoleUtil.cs [hunk 9]
```diff
@@ -39,41 +37,15 @@ public static void Write(string content, ConsoleColor color = ConsoleColor.White
...
-                    Console.CursorTop += height;
-
-                Console.WriteLine(content);
-
-                Console.ForegroundColor = ConsoleColor.White;
-                Console.BackgroundColor = ConsoleColor.Black;
-
-                //Logger.Log(content);
-            }
+            Write(content + Environment.NewLine, color, height);
```
Comment (MichalPetryka) [inline IO/ConsoleUtil.cs:48]: I'd consider using WriteLine instead of Write text + newline in case the implementation differs on some platforms

File: IO/ConsoleUtil.cs [hunk 10]
```diff
@@ -1,11 +1,10 @@
 ﻿using System;
+using LocalAdmin.V2.IO.Logging;
 
 namespace LocalAdmin.V2.IO
 {
     public static class ConsoleUtil
     {
-        private static readonly char[] ToTrim = { '\n', '\r' };
```
Comment (MichalPetryka) [inline IO/ConsoleUtil.cs:7]: Why did you remove this?

File: Core/LocalAdmin.cs [hunk 11]
```diff
@@ -305,10 +309,24 @@ private void RunScpsl(ushort port)
...
 
                 gameProcess = Process.Start(startInfo);
+
+                gameProcess.OutputDataReceived += (sender, args) =>
+                {
+                    Logger.LogLine($"[SCPSL] {args.Data}");
+                };
+                gameProcess.BeginOutputReadLine();
+
+                gameProcess.Exited += (sender, args) =>
```
Comment (MichalPetryka) [inline Core/LocalAdmin.cs:324]: Afaik LocalAdmin already handles process exit, what's the point of this?

File: Core/LocalAdmin.cs [hunk 12]
```diff
@@ -42,7 +43,7 @@ public sealed class LocalAdmin
 
         public void Start(string[] args)
         {
-            Console.Title = $"LocalAdmin v. {VersionString}";
+            Console.Title = $"LocalAdmin {VersionString}";
```
Comment (MichalPetryka) [inline Core/LocalAdmin.cs:46]: This change doesn't seem relevant to this PR.

File: Core/LocalAdmin.cs [hunk 13]
```diff
@@ -305,10 +309,24 @@ private void RunScpsl(ushort port)
                 {
                     FileName = scpslExecutable,
                     Arguments = $"-batchmode -nographics -nodedicateddelete -port{port} -console{server!.ConsolePort} -id{Process.GetCurrentProcess().Id}",
-                    CreateNoWindow = true
+                    CreateNoWindow = true,
+                    RedirectStandardOutput = true
```
[the remark this answers is not in the document]
Comment (js6pak) [reply Core/LocalAdmin.cs:313]: UseShellExecute is false by default on .net core
https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.processstartinfo.useshellexecute?view=netcore-3.1#property-value

File: Core/LocalAdmin.cs (hunk 11 above)
Comment (js6pak) [reply Core/LocalAdmin.cs:324]: LocalAdmin handles only exit of its own process not SCP:SL, so if game crashes you don't know that until you type any command.

File: IO/ConsoleUtil.cs (hunk 9 above)
Comment (js6pak) [reply IO/ConsoleUtil.cs:48]: Look at implementation that was commented before, I didn't want to make it redundant, should I use bool parameter on end of Write method or what?

File: IO/ConsoleUtil.cs (hunk 10 above)
Comment (js6pak) [reply IO/ConsoleUtil.cs:7]: To make ^ work, and it was useless anyway, blocking people who actually would want newlines in output.

File: Core/LocalAdmin.cs (hunk 12 above)
Comment (js6pak) [reply Core/LocalAdmin.cs:46]: Kinda it is, I was cleaning up LA output, I did same thing in other places.

File: IO/Logging/Logger.cs (hunk 2 above)
Comment (js6pak) [reply IO/Logging/Logger.cs:45]: Without flush it works weird, sometimes it just stops writing to file.

File: IO/Logging/Logger.cs (hunk 4 above)
Comment (js6pak) [reply IO/Logging/Logger.cs:17]: Why?

File: IO/ConsoleUtil.cs (hunk 10 above)
Comment (MichalPetryka) [reply IO/ConsoleUtil.cs:7]: Trim only removed newlines at the beginning and the end

File: IO/Logging/Logger.cs (hunk 2 above)
Comment (MichalPetryka) [reply IO/Logging/Logger.cs]: It's because StreamWriter isn't supposed to write instantly, it buffers data to increase performance

File: Core/LocalAdmin.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (MichalPetryka) [reply Core/LocalAdmin.cs:318]: It's better to explictly set it to false

File: IO/Logging/Logger.cs (hunk 2 above)
Comment (js6pak) [reply IO/Logging/Logger.cs]: It was because process exited wuth Enviroment#Exit and it didn't have chance to flush on its own

File: IO/ConsoleUtil.cs (hunk 10 above)
Comment (js6pak) [reply IO/ConsoleUtil.cs:7]: I know.

File: Core/LocalAdmin.cs [hunk 14]
```diff
@@ -310,11 +313,31 @@ private void RunScpsl(ushort port)
                 var startInfo = new ProcessStartInfo
                 {
                     FileName = scpslExecutable,
-                    Arguments = $"-batchmode -nographics -nodedicateddelete -port{port} -console{server!.ConsolePort} -id{Process.GetCurrentProcess().Id} {gameArguments}",
-                    CreateNoWindow = true
+                    Arguments = $"-batchmode -nographics -nodedicateddelete -port{port} -console{server!.ConsolePort} -id{Process.GetCurrentProcess().Id}",
```
[the remark this answers is not in the document]
Comment (js6pak) [reply Core/LocalAdmin.cs:316]: Probably my mistake while merging

File: IO/Logging/Logger.cs [hunk 15]
```diff
@@ -1,29 +1,61 @@
...
     {
-        public static void Log(string text)
+        public static StreamWriter? Writer { get; private set; }
```
Comment (MichalPetryka) [inline IO/Logging/Logger.cs:9]: This should be private

File: IO/Logging/Logger.cs [hunk 16]
```diff
@@ -1,29 +1,61 @@
...
+        private static object _lock = new object();
```
Comment (MichalPetryka) [inline IO/Logging/Logger.cs:10]: This should be readonly

File: IO/Logging/Logger.cs [hunk 17]
```diff
@@ -1,29 +1,61 @@
...
             }
+
+            Writer = File.AppendText(filePath);
         }
 
-        public static void Log(object obj)
+        public static void Dispose()
         {
-            if (!File.Exists("localadmin.log")) File.Create("localadmin.log");
+            Writer?.Flush();
```
Comment (MichalPetryka) [inline IO/Logging/Logger.cs:39]: Dispose automatically flushes

File: IO/Logging/Logger.cs (hunk 4 above)
Comment (MichalPetryka) [reply IO/Logging/Logger.cs:18]: Just compress logs on creation instead

File: IO/Logging/Logger.cs (hunk 16 above)
Comment (js6pak) [reply IO/Logging/Logger.cs:10]: I copied it from https://github.com/northwood-studios/LocalAdmin-V2/blob/0b0991dab9aa3cb7e01c569052173eb815568a8c/IO/ConsoleUtil.cs#L9 but ok

Comment (js6pak) [reply IO/Logging/Logger.cs:10]: What should be name (especially casing) of it then? :/

File: IO/Logging/Logger.cs (hunk 4 above)
Comment (js6pak) [reply IO/Logging/Logger.cs:18]: What do you mean?

File: IO/Logging/Logger.cs (hunk 16 above)
Comment (MichalPetryka) [reply IO/Logging/Logger.cs:10]: Ask @zabszk about naming conventions

Comment (Rnen) [reply IO/Logging/Logger.cs:10]: Also, isn't a StreamWriter nullable by default?

File: IO/Logging/Logger.cs (hunk 15 above)
Comment (Rnen) [reply IO/Logging/Logger.cs:9]: Also, isn't a StreamWriter nullable by default?

Comment (js6pak) [reply IO/Logging/Logger.cs:9]: @Rnen https://github.com/northwood-studios/LocalAdmin-V2/blob/0b0991dab9aa3cb7e01c569052173eb815568a8c/LocalAdmin%20V2.csproj#L7

**Comment by zabszk** [comment]
Implemented almost entirely in #15 and #16.
